core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -375,1083 +380,674 @@ public boolean isPreloaded() {
+                     .flatMapCompletable(entry -> Completable.fromCompletionStage(entry.getKey().write(segment, marshalledEntry)));
+            },
+            this::releaseReadLock
+      ).toCompletionStage(null);
    }
 
-   @Override
-   public CompletionStage<Void> deleteBatchFromAllNonTxStores(Iterable<Object> keys,
-         Predicate<? super StoreConfiguration> predicate, long flags) {
-      if (!keys.iterator().hasNext())
-         return CompletableFutures.completedNull();
-
-      return runOnPersistenceExAndContinue(traceId -> {
-         acquireReadLock();
-         try {
-            checkStoreAvailability();
-            if (trace) {
-               log.tracef("Continuing delete batch for id %d", traceId);
-            }
-            nonTxWriters.stream()
-                  .filter(writer -> predicate.test(getStoreConfig(writer)))
-                  .forEach(writer -> writer.deleteBatch(keys));
-         } finally {
-            releaseReadLock();
-         }
-      }, "Submitting persistence async operation of id %d to write a batch");
+   private boolean shouldWrite(StoreStatus storeStatus, Predicate<? super StoreConfiguration> userPredicate, long flags) {
+      return !storeStatus.characteristics.contains(Characteristic.READ_ONLY)
+            && userPredicate.test(storeStatus.config)
+            && !storeStatus.store.ignoreCommandWithFlags(flags);
    }
 
    @Override
    public CompletionStage<Void> prepareAllTxStores(Transaction transaction, BatchModification batchModification,
          Predicate<? super StoreConfiguration> predicate) throws PersistenceException {
-      return runOnPersistenceExAndContinue(traceId -> {
-         acquireReadLock();
-         try {
-            checkStoreAvailability();
-            if (trace) {
-               log.tracef("Continuing prepare batch for id %d", traceId);
-            }
-            for (CacheWriter writer : txWriters) {
-               if (predicate.test(getStoreConfig(writer)) || configuration.clustering().cacheMode().equals(CacheMode.LOCAL)) {
-                  TransactionalCacheWriter txWriter = (TransactionalCacheWriter) undelegate(writer);
-                  txWriter.prepareWithModifications(transaction, batchModification);
+      return Completable.using(
+            this::acquireReadLock,
+            ignore -> {
+               checkStoreAvailability();
+               if (trace) {
+                  log.tracef("Preparing batch for store: %s on transaction %s", batchModification, transaction);
                }
-            }
-         } finally {
-            releaseReadLock();
-         }
-      }, "Preparing all tx stores for id %d");
+               return Flowable.fromIterable(stores.entrySet())
+                     .filter(entry ->
+                           !entry.getValue().characteristics.contains(Characteristic.READ_ONLY)
+                                 && predicate.test(entry.getValue().config))
```

Comment:
Should we be testing against `shouldWrite` here? Or should `ignoreCommandWithFlags` be ignored for transactions?

If so, `commitAllTxStores` and `rollbackAllTxStores` also needs updating.

Reply:
I had debated about transactions. This whole thing still rubs me the wrong way. I really don't like it.

Reply:
It being the way store writes are skipped with rolling upgrade and remote store.

Reply:
This will be further handled in https://issues.redhat.com/browse/ISPN-11859